### PR TITLE
Made serial connection class to handle serial read / write and init.

### PR DESCRIPTION
**Pull request by Murray2015**
Hi @mhagdorn - you unfortunately now get to see how poor my OOP is! 

The rest of the classes aren't hooked up so this won't be working and shouldn't be merged yet (I'm sure this is obvious!), but I figure that it's easier to read a smaller amount of code as I go along, than a single huge PR at the end. 

I've created a class to hold the serial connection info, start the connection and implemented a method to hold the boilerplate for read/write calls. Please do point out any obvious bugs or mistakes. There may also be more pythonic ways of doing things, any pointers would be appreciated.

### CODE REVIEW

File: piccolo3/server/PiccoloCoolboxControl.py [hunk 1]
```diff
@@ -26,11 +26,91 @@
...
 import logging
 
 from random import randint
 
 
+class PiccoloSerialConnection(PiccoloNamedComponent):
+    """Manage serial connection for controlling and managing the coolbox"""
+
+    def __init__(self, serial_port="/dev/ttyUSB0"):
+        self.verbose = False
```
Comment (mhagdorn) [inline piccolo3/server/PiccoloCoolboxControl.py:40]: I wouldn't bother with the verbose flag - just use the logging tools. i tend to have these print statements in during the development. having said that, print doesn't work very well, so I tend to use logging

File: piccolo3/server/PiccoloCoolboxControl.py [hunk 2]
```diff
@@ -26,11 +26,91 @@
...
+        self.ser = None
+        self.initialize_serial()
+        self.initialise_coolbox()
+
+    def initialize_serial(self):
+        try:
+            self.ser = serial.Serial(
+                port=self.serial_port,
+                timeout=1,
+                baudrate=115200,
```
Comment (mhagdorn) [inline piccolo3/server/PiccoloCoolboxControl.py:57]: I wonder if it is worth making this configurable

File: piccolo3/server/PiccoloCoolboxControl.py [hunk 3]
```diff
@@ -26,11 +26,91 @@
...
+                self.ser.write(cmd_str.encode())
+                cmd_str = "$RW\r\n"
+                self.ser.write(cmd_str.encode())
+                if self.verbose:
+                    print("Coolbox initiallised")
+            self.ser.close()
+            self.log.info("Successfully initialized coolbox.")
+        except Exception as e:
+            self.log.debug("Couldn't initialize coolbox.")
+            self.log.debug(e)
```
Comment (mhagdorn) [inline piccolo3/server/PiccoloCoolboxControl.py]: this should be logged as an error

File: piccolo3/server/PiccoloCoolboxControl.py [hunk 4]
```diff
@@ -26,11 +26,91 @@
...
+        self.ser.write(cmd_str.encode())
+        await asyncio.sleep(self.tsleep)
+        serial_data = self.ser.readline()
+        serial_data = self.ser.readline()
+        serial_data = serial_data.decode()
+        serial_data = struct.unpack('!f', bytes.fromhex(serial_data))[0]
+        if self.verbose:
+            print(verbose_message + " : command string sent was " +
+                  cmd_str + " : data recieved back was " + str(serial_data) + "\n")
+        return data
```
Comment (mhagdorn) [inline piccolo3/server/PiccoloCoolboxControl.py]: I would use a lock to manage access to the serial port. you can then use a resource manager: https://docs.python.org/3/library/asyncio-sync.html#lock

**Review by mhagdorn — COMMENTED** [review]
main comment - use a lock to access serial port.
I tend to use private class attribs and define a read-only accessor, eg
```
def __init__(self):
   ...
   self._serial_port = serial_port
...
@property
def serial_port(self):
    return self._serial_port
```

File: piccolo3/server/PiccoloCoolboxControl.py (hunk 1 above)
Comment (Murray2015) [reply piccolo3/server/PiccoloCoolboxControl.py:40]: Good point. I'll leave them in until everything is working as like you say they will be useful for debugging, but will remove them before merging coolbox with main

File: piccolo3/server/PiccoloCoolboxControl.py (hunk 2 above)
Comment (Murray2015) [reply piccolo3/server/PiccoloCoolboxControl.py:57]: Possibly a good idea - I'd considered it myself. I came down on the side of not at this stage, but should be fairly easy to add in the future if needed.

File: piccolo3/server/PiccoloCoolboxControl.py (hunk 3 above)
Comment (Murray2015) [reply piccolo3/server/PiccoloCoolboxControl.py]: Good spot, I've changed throughout the code.

File: piccolo3/server/PiccoloCoolboxControl.py (hunk 4 above)
Comment (Murray2015) [reply piccolo3/server/PiccoloCoolboxControl.py]: This lock is exactly what we were looking for! That link looks fairly easy to implement for the lock, but how and where do we implement the resource manager? I've had a google and there seem to be a lot of variations of python resource managers...

Comment (mhagdorn) [reply piccolo3/server/PiccoloCoolboxControl.py]: you just use it. so in the init you create the lock
```
self._lock = asyncio.lock()
```
and then when you want to access the serial port you just do
```
with self._lock:
   self.ser.read()
```

Comment (Murray2015) [reply piccolo3/server/PiccoloCoolboxControl.py]: Ah I get you - simpler than expected! I'll implement that.

Comment (mhagdorn) [reply piccolo3/server/PiccoloCoolboxControl.py]: it's python ;-)

Comment (mhagdorn) [reply piccolo3/server/PiccoloCoolboxControl.py]: you might want to add a timeout when you are trying to acquire the lock. I guess you will also need to use timeouts for reading/writing the serial port. I suspect you can also use a resource manager when communicating with the serial port, eg
```
with serial.open(...) as ser:
    data = ser.readline()
```

Comment (Murray2015) [reply piccolo3/server/PiccoloCoolboxControl.py]: I've implemented the lock and pushed to this PR, hopefully is as simple as it looks! I suspect you're right about being able to use a resource manager with statement on serial.open. According to the docs you definitely can with serial.Serial(). I've made a mental note to see if that can be easily refactored once the code is in a testable state. 

Couldn't find much about timeouts for the asyncio lock, looks like it doesn't take a timeout param.

**Pull request merged**